### ISSUE TITLE
Unable to support more than one simultaneous connection

### ISSUE THREAD

**Issue by apavlo** [cmu-db/terrier#731]
If I run the fuzzer in one thread and then connect to the DBMS in another thread, then the first thread gets disconnected. There is no error on the server side and no useful error message in the client.

A even more simple to reproduce is to try connecting to the same server instance with separate psql instances. It doesn't always happen but once it does, only one client can connect at a time.

**Server**
```
$ ./debug/terrier 
[2020-02-01 14:31:46.082] [network_logger] [info] Listening on port 15721
```

**Client 1**
```
$ psql -h localhost -p 15721 -d terrier
psql (11.5 (Ubuntu 11.5-1), server 9.5devel)
Type "help" for help.

terrier=#
```

**Client 2**
```
$ psql -h localhost -p 15721 -d terrier
psql: server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
```

**Comment by lmwnshn** [comment]
From the meeting, it looks like we have the following bug.

While no psql client has disconnected yet, you can keep adding new clients, no problem.
However, once the initial **set** of psql clients have disconnected, you can only connect one client at a time.

I suspect the network state machine just goes to the wrong state when you go from "have >0 clients" to "no clients".

**Comment by apavlo** [comment]
Fixed in #760
